solutions/balkan-11-timeismoney.mdx

```diff
@@ -0,0 +1,180 @@
+
+At the beginning, we choose two points on the lower convex hull, $A$ and $B$, as our starting points for our search. $A$ should be the point where we minimize $T$ and $B$ where we minimize $C$. We then want to find a point $P$ which is left-most from line $AB$. The area of the triangle $ABP$ is proportional to the distance between the line $AB$ and the point $P$, given $|AB|$ is constant. Therefore, we only have to find out the triangle $ABP$ with maximum distance. Using the cross product, we get $-2A_{ABP} = AB \times AP = (x_B - x_A)(y_P - y_A) - (y_B - y_A)(x_P - x_A) = y_P(x_B - x_A) - x_P(y_B - y_A) + x_A(y_B - y_A) - y_A(x_B - x_A)$. Note that $x_A(y_B - y_A) - y_A(x_B - x_A)$ is constant and can be ignored. Thus, $-A_{ABP}$ is proportional to $y_P(x_B - x_A) - x_P(y_B - y_A)$. As our goal was to maximize $A_{ABP}$, we should here minimize $y_P(x_B - x_A) - x_P(y_B - y_A)$ instead. This can be done by taking this value as the weight of our MST algorithm and then build the MST. The result would be our new point $P$. If $P$ is on the right side of vector $AB$ (thus on the left side of line $AB$), it is also on the convex hull. We then use Divide-and-Conquer and solve the same problem with either points $A$, $P$ or $P$, $B$. For each of these points, we update the current best solution if the new point is better. Otherwise, if $P$ is on the right side of line $AB$, we do not have to search further with $P$ since it is not on the lower convex hull anymore.
+
+Since $T$ and $C$ are positive integers, the lower convex hull between $A$ and $B$ consists of $k \leq (N-1) \cdot t_{max} = 50745$ points. For each of those $k$ points, we run Kruskal's algorithm in $\mathcal{O}(M \log M)$. Therefore, the total time complexity is $\mathcal{O}(kM \log M)$. This suffices for the given test cases. 
+
+<LanguageSection>
+<CPPSection>
+## Implementation
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long int;
+using pii = pair<int, int>;
+// [town-x, town-y, time, cost, id]
+using Edge = array<int, 5>;
+
+struct DSU {
+	vector<int> g;
+	DSU(int n) : g(n, -1) {}
+	int find(int a) {
+		if (g[a] < 0)
+			return a;
+		else
+			return g[a] = find(g[a]);
+	}
+	/**
+	 * If a and b are already in the same set, return false and do nothing.
+	 * Otherwise, merge set a and b into one set and return true.
+	 */
+	bool unite(int a, int b) {
+		a = find(a);
+		b = find(b);
+		if (a == b) return false;
+		if (g[a] > g[b]) swap(a, b);
+		g[a] += g[b];
+		g[b] = a;
+		return true;
+	}
+};
+
+/**
+ * Kruskal's Algorithm to find a MST with given edges and weights. Returns the
+ * time and cost as pair<int,int> as well as ids of edges of the MST as
+ * vector<int>.
+ */
+pair<pii, vector<int>> kruskal(int n, vector<Edge> &edges,
+										vector<int> &weights) {
```

Comment:
indentation sus

Reply:
well it looks sus on the website otherwise. Any idea what can I do?